Modify from_pandas_dataframe to accept non-square inputs

rnelsonchem opened the pull request:
First PR, so let me say that NetworkX has been very useful in the past. Thanks!

I was looking at the dev docs and was happy to see that there is some work being done to handle Pandas DataFrames as inputs for graph creation (from_pandas_dataframe). Unfortunately (for me), the source looks to be a very thin wrapper around from_numpy_matrix. This doesn't work for the DataFrames I create, which are sparse and not square (Example below). This PR modifies from_pandas_dataframe in a way that accepts non-square DFs. What I've done is to create a generator function that takes a DataFrame and returns an [ebunch](http://networkx.github.io/documentation/development/reference/glossary.html#term-ebunch). The generator is passed to the `add_edges_from` constructor. This works fine for all shapes of DataFrame, including the square case that is already covered.

Tested Python 3.4 on Linux.

There's still a couple of problems... First of all, I wrote a test that follows the from_numpy_matrix tests. Unfortunately, I wasn't working on my machine because it couldn't find `np` and `pd`. Pointers?  Second, should this have flags to ignore zero and nan? Or should they just be ignored without asking? 

Example:
I might have a connectivity DataFrame that looks like the following (just made this up):

donor|A|B|C|
---|---|---|---
accept|||
|D|2.7|0.4|0.3|
|E|0.00|0.03|0|
|F|0.00|0.7|2.1|
|G|0.52|0|1.2|

In this extreme example, there is not any overlap between the two sets and they are different lengths. I could [rebuild my DataFrame](http://stackoverflow.com/questions/21207872/construct-networkx-graph-from-pandas-dataframe) to have the same columns and indices ; however, this is a bit of extra work and creates a very large table of empty values. 

If the above Dataframe is stored in `df`, then the modified function will produce:

    G=nx.from_pandas_dataframe(df)
    print(G.edges(data=True))

[('A', 'G', {'weight': 0.52000000000000002}),
 ('A', 'D', {'weight': 2.7000000000000002}),
 ('G', 'C', {'weight': 1.2}),
 ('E', 'B', {'weight': 0.029999999999999999}),
 ('B', 'F', {'weight': 0.69999999999999996}),
 ('B', 'D', {'weight': 0.40000000000000002}),
 ('F', 'C', {'weight': 2.1000000000000001}),
 ('D', 'C', {'weight': 0.29999999999999999})]

rnelsonchem commented:
Another question, is there a preference for edge selection when creating a `Graph` with different edge weights? Here's an example from the documentation for `from_numpy_matrix`. The `weight` for (0,1) changes.

>>> import numpy
>>> A=numpy.matrix([[1,1],[2,1]])
>>> G=nx.from_numpy_matrix(A)
>>> G.edges(data=True)
[(0, 0, {'weight': 1}), (0, 1, {'weight': 2}), (1, 1, {'weight': 1})]

>>> G=nx.from_numpy_matrix(A.T)
>>> G.edges(data=True)
[(0, 0, {'weight': 1}), (0, 1, {'weight': 1}), (1, 1, {'weight': 1})]

My pandas version of this function shows the same behavior, but the (0,1) weights are inverted (ie must transpose the DF to get the same as the first example above using the same data). Does this matter?

rnelsonchem commented:
Actually, I can speed this up and simplify the code if you would allow a completely different format of the DataFrame input. For example, if you have the DataFrame as such:

Index | NodeA | NodeB |weight
---|---|---|---
0 | A | G | 0.52
1 | E | B | 0.03
etc... 

Then you can write a generator function that only needs to do a single loop through the rows of the table. (The one I've written does two nested loops.) In addition, you could pretty easily write this to accept extra columns for additional information about the edges. In this case, though, you would want to have a bunch of extra flags for the selecting node columns, weight column, etc.

I think this is going to be a more common format for people working with DataFrames in general. This is the way I extract my data. In order to get my previous attempt to work, I had to do a [pivot table](http://pandas.pydata.org/pandas-docs/stable/reshaping.html#reshaping-by-pivoting-dataframe-objects) from my original DataFrame, which generated a lot of unnecessary data.

I played around with this and got something working. Depending on what people are thinking about this change in general. I could modify my original pull request or make a new one if you'd like to see what I'm talking about.

brifordwylie commented:
Just wanted to throw in a big +1 of encouragement for this. Having a slick way of going from a Pandas dataframe to a NetworkX graph will be super useful :)

Midnighter commented:
Hi, I just made a pull request to your pull request :smile: here: https://github.com/rnelsonchem/networkx/pull/1

It goes a little further with this idea. First of all, it removes the `mask_zero` and `mask_nan` arguments. I think, this is better done outside of the function. Just pass an appropriately modified `pandas.DataFrame` to the function, e.g.:

```python
df = pd.DataFrame(...)
nx.from_pandas_dataframe(df.loc[df["junk_column"].notnull() & df["important_column"] != 0])
```

Second, you can now specify the columns of the source and target node (for a directed graph). Additionally, you can give the column name(s) that are to be used as edge attributes. At the moment the edge attributes are the same name as the column names.

This is a concept only. I did not fully test this, nor adjust the tests accordingly. What do you think?

rnelsonchem commented:
@Midnighter  Thanks for the addition. I also realized that my original code was not very Pandas-y, hence my previous post. After I discovered this function, I was trying to force my DataFrame into a shape that would work with the current function. I used a pivot table to get close, but my edge list was sparse so I was generating a gigantic full matrix of connections... And you can't include any extra metadata about the edge. 

Anyway, I've also made a version that is similar to yours. However, I like a couple things you've done here, so I'll keep a bit of it. There still needs to be a lot of extra logic and other magic incantations, but this is overall probably a better direction than my initial addition.

I'd like to hear from the core devs before I/we do too much, though. Thanks again.

rnelsonchem commented:
@Midnighter Since you brought it up, I decided to incorporate your additions. Thanks.  I think we both agree that this is a much better way to handle DataFrames in general. I had to make changes to cover more corner cases. Added some tests and fixed the documentation as well. Let's see what folks think of these modifications.

Midnighter commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,103 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+
+        length = len(edge_attr)
+        if length > 1:
+            for idx, row in df.iterrows():
```
Just pass `index=False` to iterrows and remove `idx`.

rnelsonchem replied:
Iterrows [doesn't take that kwarg](https://github.com/pydata/pandas/blob/5531341076ec4d1cded6be0ae74dc18ae630311a/pandas/core/frame.py#L544). 
Maybe we could replace these conditionals all together and bypass iterrows, since it has a lot of overhead that is unnecessary here. Let's just iterate through `df.values` directly. The core of this function would then be to find the row indices that correspond to the nodes and attributes, which wouldn't be much different than now.

Midnighter commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,103 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+        # If only one edge_attr, can't use kwarg expansion
+        elif length == 1:
+            meta = edge_attr[0]
+            for idx, row in df.iterrows():
```
Same as above.

Midnighter commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,103 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+    
+    # If no column names are given, then just return the edges.
+    else:
+        for idx, row in df.iterrows():
```
Again.

Midnighter commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,103 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+
+    if edge_attr:
+        # If all additional columns requested
+        if edge_attr == True:
```
`edge_attr is True` maybe?

Midnighter commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,103 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+        if length > 1:
+            for idx, row in df.iterrows():
+                g.add_edge(row[source], row[target], **row[edge_attr])
+        # If only one edge_attr, can't use kwarg expansion
```
Since we still made `edge_attr` a list above, we can actually use kwarg expansion.

rnelsonchem replied:
This is a pretty neat trick! That will be good to know. Thanks.

Midnighter commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,103 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+      used to retrieve attributes from the row that should be added to the
+      graph as edge attributes. If `True` is passed as the key, all of the
+      remaining columns will be added.
+
```
I found the parameter descriptions a little confusing. You're **not** using integer arguments to `source`, `target` and `edge_attr` as indeces but rather (as pandas allows) expect the column names to be integers.

rnelsonchem replied:
Okay. How about this:
```
edge_attr: str or int, iterable, True
  A valid column name (can be a string or integer) or list of column names 
  that will be used to retrieve items from the row and add them to the graph 
  as attributes. If True, all of the remaining columns will be added.
```

Midnighter commented:
OK, my bad.

It's fairly easy.

```python
from operator import itemgetter

src_i = df.columns.index(source)
tar_i = df.columns.index(target)
indeces = [df.columns.index(attr) for attr in edge_attr]
get_attr = itemgetter(indeces)

for row in df.itertuples(index=False):
    g.add_edge(row[src_i], row[tar_i], **{name: attr for (name, attr) in zip(edge_attr, get_attr(row))})
```

I'll test which method is faster. It might be worth keeping the overhead of iterating over row `Series`.

rnelsonchem commented:
@Midnighter I'm not sure why we would want to use `itertuples` in this case either. I disagree that we ultimately want to have this function change the column names. I think it is easy enough to change the DataFrame names before passing it here. You made the point earlier that we should expect the users to check for zeros and NaNs. However, I could be convinced otherwise. 

I was thinking of just iterating over `DataFrame.values` directly, which simplifies things a bit:
```
src_i = df.columns.get_loc(source)
tar_i = df.columns.get_loc(target)
edge_i = [(i, df.columns.get_loc(i)) for i in edge_attr]
for row in df.values:
    G1.add_edge(row[src_i], row[tar_i], {i:row[j] for i, j in edge_i})
```

Anyway, I guess I beat you to [the performance testing](http://nbviewer.ipython.org/gist/rnelsonchem/3b6b7277501cdc413c3d) ;) As I suspected, `iterrows` has terrible performance. Looping directly over `values` is slightly faster, but not much to write home about. At this point, it is mostly a choice of style.

Midnighter commented:
> I disagree that we ultimately want to have this function change the column names. I think it is easy enough to change the DataFrame names before passing it here. You made the point earlier that we should expect the users to check for zeros and NaNs.

I completely agree with you. The data frame should be prepared beforehand.

Looks good from my point of view.

ysitu commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,103 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+    >>> G.edges(data=True)
+    [('E','C',{'cost':9,'weight':10}),
+    ('B','A',{'cost':1,'weight':7}),
+    ('A','D',{'cost':7,'weight':4})]
```
Doctest is failing here.

rnelsonchem commented:
@ysitu I fixed those doc string issues. I don't know how to fix the import issue in the tests I wrote. The tests themselves pass fine when Numpy and Pandas are imported at the top of the script. However, having the imports in setupClass doesn't work on my computer either. It seems to work fine for some of the test systems, though... 

Anyway, I think I've incorporated all the changes based on the notes/discussion so far. I'm happy to make other modifications.

rnelsonchem commented:
Can anyone give me a pointer to get the import issue fixed in my tests? I guess I don't know enough about Nose tests to force them to run the setupClass method. I followed the format of the `test_convert_numpy.py` test class, and my tests run fine on some of the setups...

chebee7i commented:
Apologies for the delay....

For the import errors, this seems to be a bit of a nose quirk. `setupClass` is run immediately to determine if any of the method tests should be ran. So if NumPy is not available, then it will skip. However and confusingly, this does not seem to imply that `np` will be globally available. In fact, it seems that only methods which match the regex for testing will have the same namespace that includes `np`. So I think you'll have to add an explicit `import numpy as np` in your `__init__` method.

Once you get that fixed, is this ready to be merged?

rnelsonchem commented:
@chebee7i Thanks for the suggestion. It seems to be working now. (I have to admit that I still don't understand why my case was not working. The Scipy tests are set up in a very similar way to my original tests, and they pass fine.)

I think this is now complete from my perspective. 

Note: The Travis-CI test case "Python 3.2 OPTIONAL_DEPS=miniconda" does not seem to be working properly. The tests appear to pass, but I don't think that Numpy, Scipy, or Pandas are installed. Maybe Continuum doesn't support these packages for Python 3.2 any longer.

chebee7i commented:
Great. If you have the patience to do a few more edits, we can get this merged. If not, let me know and I'll take care of the rest.

Could you edit:  `doc/source/reference/api_2.0.rst` and add a short comment about this PR. Also, consider editing `doc/source/reference/credits.rst` to add whatever information you feel is relevant (your personal info, acknowledgments for any support you've received, any affiliations etc.).

I had a few other comments that you'll see shortly.

chebee7i commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,101 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
-    If the numpy matrix has a single data type for each matrix entry it
-    will be converted to an appropriate Python data type.
+    target : str or int
+      A valid column name (string or iteger) for the target nodes (for the
```
Indentation for the parameter description should be 4 spaces in (according to the NumPy docstring standard, which we follow). So if you could update all the descriptions in each docstring---it looks like you need to add 2 more spaces.

chebee7i commented on networkx/convert_matrix.py:
```diff
@@ -121,51 +121,101 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+    if edge_attr:
+        # If all additional columns requested, build up a list of tuples
+        # [(name, index),...]
+        if edge_attr == True:
```
This works.  Booleans are singletons in Python, so I guess my inclination would be to do:   `if edge_attr is True` in this particular situation---where you are distinguishing an actual boolean from something which evaluates to true.

rnelsonchem commented:
@chebee7i I made the changes you suggested. It looks like this causes some merge conflicts, so I can't determine if my changes inadvertently broke something. Hope all is well.

chebee7i commented:
Ok great, I will merge this manually.

rnelsonchem commented:
Great! This will make a few things a bit easier for me. Thanks for
accepting my changes.

On Thu, Apr 30, 2015 at 1:40 PM, chebee7i <notifications@github.com> wrote:

> Merged #1340 <https://github.com/networkx/networkx/pull/1340>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/networkx/networkx/pull/1340#event-294299053>.
>
